python/xir/parser.py

```diff
@@ -191,18 +191,20 @@ def gatestmt(self, args):
         a gate declaration."""
 
         adjoint = False
-        ctrl_wires = tuple()
+        ctrl_wires = []
 
         while args[0] in ("adjoint", "ctrl"):
             a = args.pop(0)
             if a == "adjoint":
                 adjoint = not adjoint
             elif a == "ctrl":
-                ctrl_wires += args.pop(0)[1]
+                for cw in args.pop(0)[1]:
+                    if cw not in ctrl_wires:
+                        ctrl_wires.append(cw)
```

Comment:
If `ctrl_wires` was a Python [`set`](https://docs.python.org/3/library/stdtypes.html#set):
```suggestion
                ctrl_wires.update(args.pop(0)[1])
```

Reply:
True, but I was deliberately keeping the order of the wires here. I'm not sure that's necessary though, since the order shouldn't matter (I think). 🤔

Reply:
It shouldn't matter though, right?  My understanding is that a gate with control qubits `a`, `b`, and `c` applies the gate iff `a`,  `b`, and `c` are `|1>`.  As a result, permuting `a`, `b`, and `c` makes no functional difference.

Reply:
That's true, I've updated it to use a set instead.